server/src/main/java/org/elasticsearch/common/util/Maps.java

```diff
@@ -74,22 +78,31 @@
     public static <K, V> Map<K, V> copyMapWithRemovedEntry(final Map<K, V> map, final K key) {
         Objects.requireNonNull(map);
         Objects.requireNonNull(key);
-        assertImmutableMap(map, key, map.get(key));
+        assert assertImmutableMap(map, key, map.get(key));
         return map.entrySet().stream().filter(k -> key.equals(k.getKey()) == false)
             .collect(Collectors.toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue));
     }
 
-    private static <K, V> void assertImmutableMap(final Map<K, V> map, final K key, final V value) {
-        if (Assertions.ENABLED) {
-            boolean immutable;
-            try {
-                map.put(key, value);
-                immutable = false;
-            } catch (final UnsupportedOperationException e) {
-                immutable = true;
-            }
-            assert immutable : "expected an immutable map but was [" + map.getClass() + "]";
+    // map classes that are known to be immutable, used to speed up immutability check in #assertImmutableMap
+    private static final Set<Class<?>> IMMUTABLE_MAP_CLASSES = Set.of(
+            Collections.emptyMap().getClass(),
+            Collections.unmodifiableMap(new HashMap<>()).getClass(),
+            Map.of().getClass(),
+            Map.of("a", "b").getClass()
+    );
+
+    private static <K, V> boolean assertImmutableMap(final Map<K, V> map, final K key, final V value) {
+        // check in the known immutable classes map first, most of the time we don't need to actually do the put and throw which is slow to
+        // the point of visibly slowing down internal cluster tests without this short-cut
+        if (IMMUTABLE_MAP_CLASSES.contains(map.getClass())) {
+            return true;
+        }
+        try {
+            map.put(key, value);
+            assert false;
```

Comment:
Should this be return false instead of assert?

Reply:
++ sounds good